mlflow/data/pandas_dataset.py

```diff
@@ -0,0 +1,149 @@
+        trimmed_df = df.head(MAX_ROWS)
+        # hash trimmed dataframe contents
+        md5 = hashlib.md5(pd.util.hash_pandas_object(trimmed_df).values)
+        # hash dataframe dimensions
+        n_rows = len(df)
+        md5.update(np.int64(n_rows))
+        # hash column names
+        columns = df.columns
+        for x in columns:
+            md5.update(x.encode())
+        return md5.hexdigest()[:8]
+
+    def _to_dict(self, base_dict: Dict[str, str]) -> Dict[str, str]:
+        """
+        :param base_dict: A string dictionary of base information about the
+                          dataset, including: name, digest, source, and source
+                          type.
+        :return: A string dictionary containing the following fields: name,
+                 digest, source, source type, schema (optional), profile
+                 (optional).
+        """
+        base_dict.update(
+            {
+                "schema": json.dumps({"mlflow_colspec": self.schema.to_dict()}),
+                "profile": json.dumps(self.profile),
+            }
+        )
+        return base_dict
+
+    @property
+    def source(self) -> FileSystemDatasetSource:
+        return self._source
+
+    @property
+    def df(self) -> pd.DataFrame:
+        return self._df
+
+    @property
+    def profile(self) -> Optional[Any]:
+        """
+        A profile of the dataset. May be None if no profile is available.
+        """
+        return {
+            "num_rows": len(self._df),
+            "num_elements": int(self._df.size),
+        }
+
+    @cached_property
+    def schema(self) -> Schema:
+        """
+        An MLflow ColSpec schema representing the columnar dataset
+        """
+        return _infer_schema(self._df)
+
+    def to_pyfunc(self) -> PyFuncInputsOutputs:
+        """
+        Converts the dataset to a collection of pyfunc inputs and outputs for model
+        evaluation. Required for use with mlflow.evaluate().
+        May not be implemented by all datasets.
+        """
+        if self._targets:
+            inputs = self._df.drop(columns=[self._targets])
+            outputs = self._df[self._targets]
+            return PyFuncInputsOutputs(inputs, outputs)
+        else:
+            return PyFuncInputsOutputs(self._df)
+
+
+def from_pandas(
+    df: pd.DataFrame,
+    source: Union[str, DatasetSource],
+    name: Optional[str] = None,
+    digest: Optional[str] = None,
+) -> PandasDataset:
+    """
+    :param df: A Pandas DataFrame
+    :param source: The source from which the DataFrame was derived, e.g. a filesystem
+                    path, an S3 URI, an HTTPS URL, a delta table name with version, or
+                    spark table etc. Attempting to use other source types will throw.
+    :param targets: An optional target column name or list of target column names for
+                    supervised training. The columns must be present in the dataframe
+                    (`df`).
```

Comment:
Looks like we're missing the `targets` param from the `from_pandas` method. Can we add it and also add some tests?

The `targets` param of the `PandasDataset` constructor is of type `str`, implying that the target column name is just a single string instead of a list of strings. I think that's a good place to start, so can we remove the "list of target column names" from the documentation?